### ISSUE TITLE
C: printf(:call <SNR>209_snippets_expand_or_jump('    printf',11) when `tab` twice 

### ISSUE THREAD

**Issue by Gabirel** [wsdjeg/SpaceVim#356]
## Expected behavior
Errors or something I don't really know just happened.
Especially `Tab` twice

For example: 
`:call <SNR>209_snippets_expand_or_jump('    printf',11)`

## Environment Information
- OS: Arch Linux
- vim version: 8.0

## The reproduce ways from Vim starting (Required!)
1. `vim test.c`, and code like this(`print`, not `printf`)
![image](https://cloud.githubusercontent.com/assets/12933851/24308792/5f0a738e-1104-11e7-8a02-2376050f51e6.png)

2. press `Tab` firstly, and the bar should become grey or so
![image](https://cloud.githubusercontent.com/assets/12933851/24308860/9b95466c-1104-11e7-9875-ca54d0dd4523.png)

3. press `Tab` secondly, and the error shows up
![image](https://cloud.githubusercontent.com/assets/12933851/24308969/013e0bac-1105-11e7-9e92-725759cd60de.png)

## Output of the ":message" command, and ":echo SpaceVim#logger#viewLog()"
Nothing special here

**Comment by wsdjeg** [comment]
@Shougo I can reproduce it with
```viml
augroup neosnippet_complete_done
  autocmd!
  autocmd CompleteDone * call s:my_complete_done()
augroup END
function! s:my_complete_done() abort "{{{
  if !empty(get(v:,'completed_item',''))
    let snippet = neosnippet#parser#_get_completed_snippet(v:completed_item,neosnippet#util#get_cur_text(), neosnippet#util#get_next_text())
    if snippet ==# ''
      return
    endif
    let [cur_text, col] = neosnippet#mappings#_pre_trigger()[0:1]
    call neosnippet#view#_insert(snippet, {}, cur_text, col)
  endif
endfunction"}}}
```

**Comment by wsdjeg** [comment]
I think it is due to there is a ``` \n ``` in the snippet.

**Comment by Shougo** [comment]
I have fixed the problem.  Please use the latest of neosnippet.

**Comment by wsdjeg** [comment]
Thanks, will have a try.

**Comment by wsdjeg** [comment]
it works well when popmenu is closed, but if it is open, and I still can reproduce this issue.
https://asciinema.org/a/6xeu7zt3liepwseli5n7ki2m5

when I use `tab`, I get this issue.
here is my tab mapping
```viml
function! SpaceVim#mapping#tab() abort
  if getline('.')[col('.')-2] ==# '{'&& pumvisible()
    return "\<C-n>"
  endif
  if index(g:spacevim_plugin_groups, 'autocomplete') != -1
    if neosnippet#expandable() && getline('.')[col('.')-2] ==# '(' && !pumvisible()
      return "\<Plug>(neosnippet_expand)"
    elseif neosnippet#jumpable()
          \ && getline('.')[col('.')-2] ==# '(' && !pumvisible() 
          \ && !neosnippet#expandable()
      return "\<plug>(neosnippet_jump)"
    elseif neosnippet#expandable_or_jumpable() && getline('.')[col('.')-2] !=#'('
      return "\<plug>(neosnippet_expand_or_jump)"
    elseif pumvisible()
      return "\<C-n>"
    else
      return "\<tab>"
    endif
  elseif pumvisible()
    return "\<C-n>"
  else
    return "\<tab>"
  endif
endfunction
```

**Comment by Shougo** [comment]
Unfortunately, it is out of support.
Because you use hacked CompleteDone event.

**Comment by wsdjeg** [comment]
As I know neosnippet remove completeDone support, So I use this feature. Does neosnippet support completeDone feature now?

**Comment by wsdjeg** [comment]
@Shougo I will create mini vimrc for this issue, and I found this only happend in neovim.

**Comment by wsdjeg** [comment]
@Shougo I found this is issue with deoplete-clang2, and here is mini vimrc

```viml
set rtp+=~/.cache/vimfiles/repos/github.com/Shougo/neosnippet.vim
set rtp+=~/.cache/vimfiles/repos/github.com/Shougo/neosnippet-snippets
set rtp+=~/.cache/vimfiles/repos/github.com/Shougo/deoplete.nvim
set rtp+=~/.cache/vimfiles/repos/github.com/tweekmonster/deoplete-clang2
let g:deoplete#enable_at_startup = get(g:, 'deoplete#enable_at_startup', 1)

imap <expr><Tab>
            \ neosnippet#expandable_or_jumpable() ?
            \ "\<Plug>(neosnippet_expand_or_jump)" : "\<Tab>"
```

reproduce step:
1. open vim via vim -u test.vim
2. :e test.c
3. type i to insert mode, then type into `main<tab>`

**Comment by wsdjeg** [comment]
you will see 
```
main<Plug>(neosnippet_expand_or_jump)
```

**Comment by wsdjeg** [comment]
clossing, and let's move to https://github.com/tweekmonster/deoplete-clang2/issues/12

**Comment by Shougo** [comment]
OK.
